apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,309 @@
+    return existed;
+  }
+
+  function addPlace(place) {
+    // 1. get place by url(id of DataStore)
+    // 2.A merge the existed one and new one if it's an existed one,
+    //     and update the places.
+    // 2.B Add a new record with RevisionId.
+
+    var id = place.url;
+    var revisionId;
+    return _ensureStore().then(placesStore => {
+      revisionId = placesStore.revisionId;
+      return placesStore.get(id);
+    }).then(existedPlace => {
+      if (existedPlace) {
+        var newPlace = mergeRecordsToDataStore(existedPlace, place);
+        return placesStore.put(newPlace, id, revisionId);
+      } else {
+        return placesStore.add(place, id, revisionId);
+      }
+    }).catch(e => {
+      console.error(e);
+      return;
+    });
+  }
+
+  function updatePlaces(places) {
+    return new Promise(resolve => {
+      places.reduce((cur, next) => {
+        return cur.then(() => {
+          if (next.deleted) {
+            return deletePlaceByFxSyncId(next.fxsyncId);
+          } else {
+            return addPlace(next);
+          }
+        });
+      }, Promise.resolve()).then(resolve);
+    });
+  }
+
+  function traverseRecords(each, revisionId) {
+    return _ensureStore().then(placesStore => {
+      return new Promise(resolve => {
+        var cursor = placesStore.sync(revisionId);
+        runNextTask(cursor);
+
+        function runNextTask(cursor) {
+          cursor.next().then(function(task) {
+            manageTask(cursor, task);
+          });
+        }
+
+        function manageTask(cursor, task) {
+          if (task.operation === 'done') {
+            resolve(task);
+            return;
+          }
+          each(task).then(() => {
+            runNextTask(cursor);
+          });
+        }
+      });
+    });
+  }
+
+  function deletePlace(id) {
+    return _ensureStore().then(placesStore => {
+      return placesStore.remove(id);
+    });
+  }
+
+  function deletePlaceByFxSyncId(fxsyncId) {
+    var deletingTask;
+
+    function each(task) {
+      if (task.data && task.data.fxsyncId === fxsyncId) {
+        deletingTask = task;
+      }
+      return Promise.resolve();
+    }
+
+    return traverseRecords(each).then(() => {
+      if (deletingTask) {
+        return deletePlace(deletingTask.id);
+      }
+      return Promise.resolve();
+    });
+  }
+
+  return {
+    mergeRecordsToDataStore: mergeRecordsToDataStore,
+    setSyncedCollectionMtime: setSyncedCollectionMtime,
+    getSyncedCollectionMtime: getSyncedCollectionMtime,
+    updatePlaces: updatePlaces,
+    deletePlaceByFxSyncId: deletePlaceByFxSyncId
+  };
+})();
+
+DataAdapters.history = {
+/**
+    KintoCollection.list() provides a list contains all records sorted by
+  "last_modified" property which indicates the data-modified time in
+  descending order. The "last_modified" value can be recongized as which
+  records have already synchronized before. It is used to stop early and avoid
+  the redundant operations which had handled.
```

Comment:
`KintoCollection.list() provides a list containing all the remotely retrieved Firefox Sync records sorted by "last_modified" property in descending order. After each sync request we save the "last_modified" property of the last processed record so we avoid going through the same records on following operations.`